dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java

```diff
@@ -623,6 +644,70 @@ protected Map<String, DockerParameter> getCallsToDockerMap(String mainDescriptor
         return map;
     }
 
+    /**
+     * Looks for the first line like <code>nextflow.enable.dsl=2</code>, and returns the value, 2 in this
+     * example.
+     * @param fileContent
+     * @return
+     */
+    protected Optional<String> getDslVersion(String fileContent) {
+        try {
+            List<GroovySourceAST> assignmentsAst = getGroovySourceASTList(fileContent, "=");
+            final Optional<String> dslVersion = assignmentsAst.stream()
+                .filter(equalsAst -> {
+                    final AST firstDotAst = equalsAst.getFirstChild();
+                    if (isNodeText(firstDotAst, ".")) {
+                        final AST secondDotAst = firstDotAst.getFirstChild();
+                        if (isNodeText(secondDotAst, ".") && isNodeText(secondDotAst.getNextSibling(), "dsl")) {
+                            final AST nextflowAst = secondDotAst.getFirstChild();
+                            if (isNodeText(nextflowAst, "nextflow")) {
+                                final AST enableAst = nextflowAst.getNextSibling();
+                                if (isNodeText(enableAst, "enable")) {
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                    return false;
+                })
+                .map(ast -> ast.getFirstChild().getNextSibling().getText())
+                .findFirst();
+            return dslVersion;
+        } catch (IOException | TokenStreamException | RecognitionException e) {
+            LOG.warn("could not parse", e);
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Calculates the DSL version for a workflow. Given a <code>configuration</code>, that is the
+     * model of nextflow.config file, and a <code>mainScript</code>, which is the main script referenced
+     * by nextflow.config, then:
+     *
+     * <ol>
+     *     <li>If the DSL is specified in the main script, that is the version</li>
+     *     <li>If the DSL is not specified in the main script, but specified in nextflow.config, that is the version</li>
+     *     <li>If it's in neither, we can't calculate the version -- it's based on on the engine version
+     *     or a environment setting</li>
+     * </ol>
+     * @param configuration
+     * @param mainScript
+     * @return
+     */
+    private Optional<String> calculateDslVersion(final Configuration configuration, final Optional<SourceFile> mainScript) {
+        final Optional<String> dslInScript = mainScript.flatMap(sf -> getDslVersion(sf.getContent()));
+        if (dslInScript.isEmpty()) {
```

Comment:
No need to change, but if we wanted to go full-bore on the streaminess: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Optional.html#or(java.util.function.Supplier)
The resulting "one-liner" is something like:
```
return mainScript
    .flatMap(sf -> getDslVersion(sf.getContent()))
    .or(() -> Optional.ofNullable(configuration.getString("nextflow.enable.dsl", null)));
```

Reply:
Cool, I didn't actually know about `or`, and I do prefer your suggestion.